modules/core/shared/src/main/scala/io/circe/Error.scala

```diff
@@ -96,6 +123,13 @@ object DecodingFailure {
     val path = CursorOp.opsToPath(failure.history)
     s"DecodingFailure at ${path}: ${failure.message}"
   }
+
+  sealed trait Reason
+  case class MissingField(targetType: String) extends Reason
+  case class UnsupportedType(targetType: String) extends Reason
+  case class WrongTypeExpectation(expectedJsonFieldType: String, jsonValue: Json, targetType: String) extends Reason
```

Comment:
I'm wondering whether `expectedJsonFieldType` should be a `NonEmptyList[JsonType]` where JsonType is an enum.

Also instead of `targetType` which leaks implementation details, should we have a `message` instead? As a user of an API I probably have no idea what a `NonEmptyList` is (and nor should I need to know). A message like "The array must have at least 1 element" is more actionable.

Similar sentiments about IncorrectValue's `targetType` field. If we remove `targetType` field there we can remove IncorrectValue altogether and I don't think we're really losing any expressiveness here.

Some examples error cases I can think of:

* Wrong primitive type => `WrongTypeExpectation`
* Empty array for NonEmptyString => CustomReason
* `java.time.Period` where `start` cannot be after `end` => CustomReason
* Unrecognized discriminator when trying to decode to a ADT => CustomReason (`"unrecognized discriminator 'not_a_valid_one'. Possible values are ..."`)

TBH, I'm starting to doubt whether we need WrongTypeExpectation at all...

Reply:
About targetType, I'd be 100% to just remove that. The only reason I added that was to keep a sort of continuity with the former error messages (e.g. `"[A]Option[A]"` when failing to decode an `Option[A]`, "string" for `String`, etc..).

About the other points:
-`WrongTypeExpectation` has been thought to hold the failures related to wrong primitive types in the JSON
- `IncorrectValue` to hold failures due to validations (e.g. empty list when a NEL is required, refinements, etc...)
- `CustomReason` to hold all the custom messages.

 I think that we can get rid of `targetType` and collapse `IncorrectValue` to `CustomReason` without losing any expressivity, but I think that `WrongTypeExpectation` with a proper list of expected `JsonType` may still be helpful.

That said, I'm happy to apply any amendments to this regardless :).

Reply:
Hi @jatcwang ,
I just removed `targetType` and collapsed `IncorrectValue` to `CustomReason` in 6f4a601.

Do you have any suggestion about the `JsonType` enum to show the expected json type? Maybe some representation we can reuse without redefining a new one?

Reply:
Cool :) I don't think there's anything we can reuse for JsonType within the circe codebase, no. IDM leaving it as a String since all I really need is the ability to distinguish between missing or invalid fields. Let's see what others say.

I also still see UnsupportedType, are you removing it atm?